src/main/java/org/neo4j/shell/tools/imp/format/kryo/KryoWriter.java

```diff
@@ -0,0 +1,113 @@
+            reporter.update(0, 1, props);
+        }
+        // WRITE END DUMP
+        kryo.writeObject(output, KryoSerializationTypes.DUMP_END);
+    }
+
+    private void writeConstraint(Output output, ConstraintDefinition constraintDefinition) {
+        // WRITE CONSTRAINT START
+        kryo.writeObject(output, KryoSerializationTypes.CONSTRAINT);
+        // WRITE CONSTRAINT LABEL
+        kryo.writeObject(output, constraintDefinition.getLabel().name());
+        // WRITE PROPERTY KEYS
+        List<String> propertyKeys = new LinkedList<>();
+        Iterator<String> propertyKeyIterator = constraintDefinition.getPropertyKeys().iterator();
+        while (propertyKeyIterator.hasNext()) {
+            propertyKeys.add(propertyKeyIterator.next());
+        }
+        kryo.writeObject(output, propertyKeys);
+    }
+
+    private void writeIndex(Output output, IndexDefinition indexDefinition) {
+        // Constraints indices are handled when we write out constraints
+        if (!indexDefinition.isConstraintIndex()) {
+            // WRITE INDEX START
+            kryo.writeObject(output, KryoSerializationTypes.INDEX);
+            // WRITE INDEX LABEL
+            kryo.writeObject(output, indexDefinition.getLabel().name());
+            // WRITE PROPERTY KEYS
+            List<String> propertyKeys = new LinkedList<>();
+            Iterator<String> propertyKeyIterator = indexDefinition.getPropertyKeys().iterator();
+            while (propertyKeyIterator.hasNext()) {
+                propertyKeys.add(propertyKeyIterator.next());
+            }
+            kryo.writeObject(output, propertyKeys);
+        }
+    }
+
+    private int writeNode(Output output, Node node) {
+        // WRITE NODE START
+        kryo.writeObject(output, KryoSerializationTypes.NODE);
+        // WRITE ID
+        kryo.writeObject(output, node.getId());
+        // WRITE LABELS
+        List<String> labels = new LinkedList<>();
+        Iterator<Label> labelIterator = node.getLabels().iterator();
+        while (labelIterator.hasNext()) {
+            labels.add(labelIterator.next().name());
+        }
+        kryo.writeObject(output, labels);
+        // WRITE PROPERTIES
+        Map<String, Object> properties = node.getAllProperties();
+        kryo.writeObject(output, node.getAllProperties());
+        return properties.size();
+    }
+
+    private int writeRelationship(Output output, Relationship relationship) {
+        // WRITE RELATIONSHIP
+        kryo.writeObject(output, KryoSerializationTypes.RELATIONSHIP);
+        kryo.writeObject(output, relationship.getId());
+        // WRITE SOURCE
+        kryo.writeObject(output, relationship.getStartNode().getId());
+        // WRITE TARGET
+        kryo.writeObject(output, relationship.getEndNode().getId());
+        // WRITE LABEL
+        kryo.writeObject(output, relationship.getType().name());
+        // WRITE PROPERTIES
+        Map<String, Object> properties = relationship.getAllProperties();
```

Comment:
Always having a map (empty or not) saves us from writing out an extra flag indicating whether or not there is a map.  I think that the way kryo works, the difference in size is pretty much negligible in practice, but I can check that to make sure.